### DISCUSSION TITLE
Mocking the connect function of nats

### DISCUSSION THREAD

**Discussion by chethan-b-hpe** [nats-io/nats.go#1665] [Q&A] [answered]
Any one having idea on how we can mock the nats.Connect function I tried in many ways but still its not working,
sample snippet , I want to mock Connect function of nats here :
```
`	for retries = 0; retries < maxAttempts; retries++ {
		log.Infof("connecting to: %v", url)
		conn, err = natsgo.Connect(url, nats.Timeout(200*time.Millisecond))
		if err == nil {
			break
		}

		log.Errorf("Error occurred while connecting to NATS stream. Err :%v Retry attempt %d in %d seconds", err, retries+1, backoffTime/time.Second)
		backoffTime *= 2
		time.Sleep(backoffTime)
	}`
```

**Comment by Jarema** [comment, marked as answer]
There is no mock available.
Considering how rich is NATS API, it would be really hard, if not impossible, to mock everything in a reliable way, or a way that helps more than harms.

We recommend running an actual nats-server for tests. It is really lightweight and especially easy in Go, where you can run it embedded into your tests.

**Comment by cbrake** [comment]
Running the NATS server in tests works well for us:

https://github.com/simpleiot/simpleiot/blob/master/client/rule_test.go

The NATS Go test code is full of examples of this as well.
